File: Client/main/java/com/example/securechat/crypto/SignatureBase.java
package com.example.securechat.crypto;

import org.json.JSONObject;

public abstract class SignatureBase {
    protected String algorithmName;

    public SignatureBase(String algorithmName) {
        this.algorithmName = algorithmName;
    }

    // Abstract methods - must be implemented by subclasses
    public abstract KeyPair generateSignatureKeyPair();
    public abstract Signature signMessage(String message, String privateKey) throws Exception;
    public abstract boolean verifySignature(String message, Signature signature, JSONObject publicKey);

    // Optional: Validate signature
    public boolean validateSignature(Signature signature) {
        return true; // Default implementation
    }

    // Get algorithm name
    public String getAlgorithmName() {
        return algorithmName;
    }

    // Get signature size in bits
    public abstract int getSignatureSize();

    // Get signature parameters
    public abstract JSONObject getParameters() throws Exception;

    // Abstract inner class for signature - MUST be extended by subclasses
    public static abstract class Signature {
        public String algorithm;

        public Signature(String algorithm) {
            this.algorithm = algorithm;
        }

        // Abstract method - must be implemented by subclasses
        public abstract JSONObject toJSON() throws Exception;

        // Static method to create signature from JSON
        public static Signature fromJSON(JSONObject json, String algorithmType) throws Exception {
            String algorithm = json.getString("algorithm");

            // Factory method to create appropriate signature type
            switch (algorithm) {
                case "ECDSA-P192":
                    return ECDSASignature.ECDSASignatureData.fromJSON(json);
                // Add more cases for other signature types in the future
                case "ECDSA-P256":  // ✅ ADD THIS CASE
                    return ECDSASignature2.ECDSASignatureData.fromJSON(json);
                default:
                    throw new IllegalArgumentException("Unknown signature algorithm: " + algorithm);
            }
        }
    }

    // Concrete inner class for key pair
    public static class KeyPair {
        public String privateKey;
        public JSONObject publicKey;

        public KeyPair(String privateKey, JSONObject publicKey) {
            this.privateKey = privateKey;
            this.publicKey = publicKey;
        }
    }
}

File: Server/src/crypto/signatureBase.js
// ============================================
// File: src/crypto/signatureBase.js
// ============================================

export class SignatureBase {
  constructor(algorithmName) {
    if (new.target === SignatureBase) {
      throw new TypeError("Cannot construct SignatureBase instances directly");
    }
    this.algorithmName = algorithmName;
  }

  // Abstract methods - must be implemented by subclasses
  generateSignatureKeyPair() {
    throw new Error("Method 'generateSignatureKeyPair()' must be implemented");
  }

  async signMessage(message, privateKey) {
    throw new Error("Method 'signMessage()' must be implemented");
  }

  async verifySignature(message, signature, publicKey) {
    throw new Error("Method 'verifySignature()' must be implemented");
  }

  // Serialize signature to JSON-friendly format
  serializeSignature(signature) {
    throw new Error("Method 'serializeSignature()' must be implemented");
  }

  // Deserialize signature from JSON format
  deserializeSignature(data) {
    throw new Error("Method 'deserializeSignature()' must be implemented");
  }

  // Serialize public key to JSON-friendly format
  serializePublicKey(publicKey) {
    throw new Error("Method 'serializePublicKey()' must be implemented");
  }

  // Deserialize public key from JSON format
  deserializePublicKey(data) {
    throw new Error("Method 'deserializePublicKey()' must be implemented");
  }

  // Optional: Validate signature parameters
  validateSignature(signature) {
    return true; // Default implementation
  }

  // Get algorithm name
  getAlgorithmName() {
    return this.algorithmName;
  }

  // Get signature size in bits
  getSignatureSize() {
    throw new Error("Method 'getSignatureSize()' must be implemented");
  }

  // Get signature parameters as JSON
  getParameters() {
    throw new Error("Method 'getParameters()' must be implemented");
  }
}